### PR TITLE
#684: add tag checking for invalid characters

### PR DESCRIPTION
**Pull request by BrentonPoke**
This is to address a concern brought up in #684 where strings like "mad\nrid" with a \n character would bork a tag. I approached this through a set of two regular expressions:
`nameRegex = "([a-zA-Z0-9-_]+)"`
`valueRegex = "[\x00-\x7F]+"`

Names of tags I interpret as being more strict than values, which should probably be allowed the full range of ASCII characters. Current unknowns are weird preferences others might want to report in values. I personally think ASCII codes 0-127 should be sufficient for everybody's needs, but comment below.

These are applied through a utils subpackage created in the dto package. It's not intended to be used by the developer, but I wanted something slightly portable that can be used in dto classes in the future. <strike>I also added a new exception to the `InfluxDBException` wrapper so the modified tag methods in the `Point` and `BatchPoints` classes can through it to the developer to let them know what's wrong.</strike> There is a new exception, but to pass the unit tests, I've removed the throws for them.

I am new to this codebase, but am familiar with timeseries databases. Hope this is a good starting point.

### CODE REVIEW

File: src/main/java/org/influxdb/dto/utils/CheckTags.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package org.influxdb.dto.utils;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * {Purpose of This Type}.
```
Comment (majst01) [inline src/main/java/org/influxdb/dto/utils/CheckTags.java:7]: Please describe the Type here.

File: src/main/java/org/influxdb/dto/utils/CheckTags.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * {Other Notes Relating to This Type (Optional)}
+ *
+ * @author Brenton Poke
+ *
+ */
+public final class CheckTags {
+  static final String NAMEREGEX = "([a-zA-Z0-9-_]+)";
```
Comment (majst01) [inline src/main/java/org/influxdb/dto/utils/CheckTags.java:15]: Can you please point here to the official influxdb documentation where these restrictions are described and which influxdb versions are affected.

File: src/main/java/org/influxdb/dto/Point.java [hunk 3]
```diff
@@ -131,7 +135,9 @@ public Builder tag(final String tagName, final String value) {
      */
     public Builder tag(final Map<String, String> tagsToAdd) {
```
Comment (majst01) [inline src/main/java/org/influxdb/dto/Point.java:136]: Why do we have different checks in these two tag methods ?

File: src/main/java/org/influxdb/dto/Point.java [hunk 4]
```diff
@@ -116,7 +117,10 @@ private static void throwExceptionIfMissingAnnotation(final Class<?> clazz,
     public Builder tag(final String tagName, final String value) {
       Objects.requireNonNull(tagName, "tagName");
       Objects.requireNonNull(value, "value");
-      if (!tagName.isEmpty() && !value.isEmpty()) {
+      if (!tagName.isEmpty()
```
Comment (majst01) [inline src/main/java/org/influxdb/dto/Point.java:120]: Missing test coverage for the case where tag or value is nil or empty

File: src/main/java/org/influxdb/dto/utils/CheckTags.java (hunk 2 above)
Comment (BrentonPoke) [reply src/main/java/org/influxdb/dto/utils/CheckTags.java:15]: There are no restrictions described in the documentation, which is the reason newline characters weren't being handled. I asked in #684 what other restrictions there should be, no feedback. These were reasonable starting points for that discussion.

File: src/main/java/org/influxdb/dto/Point.java (hunk 3 above)
Comment (BrentonPoke) [reply src/main/java/org/influxdb/dto/Point.java:136]: That was a mistake.

File: src/main/java/org/influxdb/dto/utils/CheckTags.java (hunk 2 above)
Comment (majst01) [reply src/main/java/org/influxdb/dto/utils/CheckTags.java:15]: OK, but i don't want to add restrictions here which are guessed, this will lead to a plethora of bug reports points not able to be written because of that.

Comment (BrentonPoke) [reply src/main/java/org/influxdb/dto/utils/CheckTags.java:15]: Then I will only restrict newline and carriage return for #684, and maybe a new issue can be created if problems arise.

File: src/main/java/org/influxdb/dto/Point.java (hunk 4 above)
Comment (BrentonPoke) [reply src/main/java/org/influxdb/dto/Point.java:120]: This will be added.

**Comment by BrentonPoke** [comment]
Sorry about the commit history. Never used Codecov before, but I learned a few things.

**Comment by BrentonPoke** [comment]
Anybody available to review this one?

File: src/main/java/org/influxdb/dto/utils/CheckTags.java (hunk 1 above)
Comment (BrentonPoke) [reply src/main/java/org/influxdb/dto/utils/CheckTags.java]: This was added too.

**Review by BrentonPoke — COMMENTED** [review]
All these changes have been made.

**Comment by majst01** [comment]
TBH, 21 commits for a single change makes me nervous, and i am really unsure if i want to merge this PR.
@fmachado should at least check if this is worth merging.

**Comment by BrentonPoke** [comment]
You can squash the commits. I tried to, but it didn't work as intended.

**Comment by majst01** [comment]
I know that i can squash them, this is not my point.

**Comment by BrentonPoke** [comment]
Then why else would what I've done not be worth merging?

**Comment by majst01** [comment]
Thats why i asked @fmachado for additional review, its not useless.

**Comment by fmachado** [comment]
@majst01 I'll take a look, thanks.

@BrentonPoke could you please share with us what InfluxDB reference documentation (or source code) you used to decide for that regexp? I couldn't find anything in your comments. :(

**Comment by BrentonPoke** [comment]
> @BrentonPoke could you please share with us what InfluxDB reference documentation (or source code) you used to decide for that regexp? I couldn't find anything in your comments. :(

The first one just takes care of the issue brought up, which was that carriage return and newline are not properly ignored. The second one just makes sure that all characters are ascii values 0-127, which covers all of English and a number of symbols.
ASCII chart for values 0 through 127

**Comment by fmachado** [comment]
> The first one just takes care of the issue brought up, which was that carriage return and newline are not properly ignored. The second one just makes sure that all characters are ascii values 0-127, which covers all of English and a number of symbols.

Apologies for not being clear with my first message: it's important for us to know that new code (1) is compliant with the line protocol (e.g. how InfluxDB handles what you name here "invalid characters") and (2) we are backward compatible with existing implementation (e.g. no unchecked exception, changes on public APIs and internal behavior like filtering non ASCII if they were accepted before).

We have to guide our implementation here based on how InfluxDB works. Could you please link here the **InfluxDB documentation** you used to support your implementation?

**Comment by BrentonPoke** [comment]
The documentation [here](https://docs.influxdata.com/influxdb/v2.0/reference/syntax/line-protocol/) says no newline characters, but doesn't restrict ascii values for fields as far as I can tell. I can remove the value regular expression entirely, so is that ok?

**Comment by fmachado** [comment]
@BrentonPoke yes, as it's written there:

> Line protocol does not support the newline character \n in tag values or field values.

Would be OK for you to submit a PR just with the relevant code and changes required for this fix?

**Comment by BrentonPoke** [comment]
You mean throw away this commit history? I can change it on this PR right now. I guess I can transfer what's needed from this branch manually to a new one and make a new pull request, but I'm not sure what that solves that a squash merge couldn't.

**Comment by BrentonPoke** [comment]
Since I already did it, I'll create the new pr from the other branch I have.
